zero/type_util.py

```diff
@@ -30,32 +31,49 @@
 allowed_types = basic_types + typing_types + special_types + pydantic_types
 
 
-def verify_function_args(func: typing.Callable):
-    arg_count = func.__code__.co_argcount
+def verify_function_typing(func: typing.Callable):
+    signature = inspect.signature(func)
+    arg_count = len(signature.parameters)
+
     if arg_count > 1:
         raise ZeroException(
-            f"`{func.__name__}` has more than 1 args; RPC functions can have only one arg - msg, or no arg"
+            f"`{func.__name__}` has more than 1 args; "
+            "RPC functions can have only one arg - msg, or no arg"
         )
 
-    if arg_count == 1:
-        arg_name = func.__code__.co_varnames[0]
-        func_arg_type = typing.get_type_hints(func)
-        if arg_name not in func_arg_type:
-            raise ZeroException(f"`{func.__name__}` has no type hinting; RPC functions must have type hints")
-
+    for name, param in signature.parameters.items():
+        if param.annotation is inspect._empty:
+            raise ZeroException(
+                f"`{func.__name__}` argument `{name}` is not typed."
+            )
+        if not param.annotation in allowed_types:
+            raise ZeroException(
+                f"`{func.__name__}` argument `{name}` type is not supported."
+            )
+
+    if signature.return_annotation is inspect._empty:
+        raise ZeroException(
+            f"`{func.__name__}` has no return type hinting; "
+            "RPC functions must have type hints"
+        )
+    elif not signature.return_annotation in allowed_types:
+        raise ZeroException(
+                f"`{func.__name__}` return type is not supported."
+            )
 
-def verify_function_return(func: typing.Callable):
-    types = typing.get_type_hints(func)
-    if not types.get("return"):
-        raise ZeroException(f"`{func.__name__}` has no return type hinting; RPC functions must have type hints")
+    return signature
```

Comment:
I don't think "inspect" is a good solution. Most of the tests are failing, it cannot even compare types, as an example `list` is supported type but it fails to recognize it 
![image](https://user-images.githubusercontent.com/15931537/226208586-a552412a-0c74-4583-b290-b3036d0c36e0.png)